README.md
## Cellular Autommata for CS 557
In this project I used `Comoands` and `Brick` to implement sa simple cellular automata. I had a stretch goal of implementing several different rule sets, but did not get quite as far as I would have liked. I did get to implement a generic structure. that makes it very easy to implement new sets of rules.

The most interesting part of the project was using `Comonads` for the Cellular automata functions.
```
instance Comonad Grid where
  -- extract:: w a -> a
  extract (Grid mtx (x,y)) = getElem x y mtx
  -- extend::(w a -> b)->w a->w b
  extend f = fmap f . duplicate
  -- duplicate::w a -> w(w a) 
  duplicate (Grid mtx (x,y)) = Grid (matrix sizeX sizeY (\(w,z)->(Grid mtx (w,z)))) (x,y)
````

Then the next thing that was somewhat difficult to tackle was making a class for the CA that was generic enough to be used in any of the Brick functions in main, this would allow me to easily change between rule sets without any work. I settled on making a class for `Cell` as well as a `CA` class that contained both a `Cell` and some `Comonad`.

```
class CA w c | w->c where
  next_gen::w->c
  neighborValues::w->[c]
  start_state::w
  . . .

class Cell c where
  cellAttr::c->AttrName

```

ChangeLog.md
# Revision history for caFinal

## 0.1.0.0  -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


src/Main.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE AllowAmbiguousTypes #-}
{-# LANGUAGE ScopedTypeVariables #-}


module Main where
import Control.Comonad
import Control.Monad (forever, void)
import Control.Monad.IO.Class (liftIO)
import Control.Concurrent (threadDelay, forkIO)
import Data.Maybe (fromMaybe)

import Brick
    ( App(..), AttrMap, BrickEvent(..), EventM, Next, Widget
    , customMain, neverShowCursor
    , continue, halt
    , hLimit, vLimit, vBox, hBox
    , padRight, padLeft, padTop, padAll, Padding(..)
    , withBorderStyle
    , str
    , attrMap, withAttr, emptyWidget, AttrName, on, fg
    , (<+>)
    )
import Brick.BChan (newBChan, writeBChan)
import qualified Brick.Widgets.Border as B
import qualified Brick.Widgets.Border.Style as BS
import qualified Brick.Widgets.Center as C
import qualified Graphics.Vty as V
import Data.Sequence (Seq)
import qualified Data.Sequence as S
import Linear.V2 (V2(..))
import Lens.Micro ((^.))
import Data.Matrix


import CellularAutomata

drawGrid::CA (w c) c=>(w c)-> [Widget Name]
drawGrid grid = [withBorderStyle BS.unicodeBold
            $ B.borderWithLabel (str "Conway's Game of Life")
            $ vBox rows]
        where
            rows = [hBox $ rowCells r | r <- [1..100]] 
            rowCells  y = [drawCell x | x <- gridRow y grid]
            

drawCell :: CA w c=>c-> Widget ()
drawCell cell = withAttr (cellAtt cell) cw





cw::Widget ()
cw = str " "


aliveAttr, deadAttr::AttrName
aliveAttr = "aliveAttr"::AttrName
deadAttr = "deadAttr"::AttrName
alvAttr = "alvAttr"::AttrName
dedAttr = "dedAttr"::AttrName
dyingAttr = "dyingAttr"::AttrName

attMap::AttrMap
attMap = attrMap V.defAttr
    [
        (aliveAttr, V.red `on` V.red)   
    ,   (deadAttr, V.yellow `on` V.yellow)
    ,   (alvAttr, V.blue `on` V.blue)
    ,   (dedAttr, V.black `on` V.black)
    ,   (dyingAttr, V.white `on` V.white)
    ]


data Tick = Tick

type Name = ()

app::(Comonad w, CA (w c) c)=>App (w c) Tick Name
app = App{  appDraw = drawGrid
          , appChooseCursor = neverShowCursor
          , appHandleEvent = handleEvent
          , appStartEvent = return
          , appAttrMap = const attMap}


--handleEvent::CA w a=>w a->BrickEvent Name Tick->EventM Name (Next (w a))
handleEvent::(Comonad w, CA (w c) c)=>(w c)->BrickEvent Name Tick -> EventM Name (Next (w c))
handleEvent g (AppEvent Tick) = continue $ extend next_gen g

main :: IO ()
main = do
    chan <- newBChan 10
    forkIO $ forever $ do
        writeBChan chan Tick
        threadDelay 100
    void $ customMain (V.mkVty V.defaultConfig) (Just chan) app start_state_brian

.stack-work/ghci/b574d4d4/cabal_macros.h
/* DO NOT EDIT: This file is automatically generated by Cabal */

/* package caFinal-0.1.0.0 */
#define VERSION_caFinal "0.1.0.0"
#define MIN_VERSION_caFinal(major1,major2,minor) (\
  (major1) <  0 || \
  (major1) == 0 && (major2) <  1 || \
  (major1) == 0 && (major2) == 1 && (minor) <= 0)

/* package ansi-terminal-0.6.3.1 */
#define VERSION_ansi_terminal "0.6.3.1"
#define MIN_VERSION_ansi_terminal(major1,major2,minor) (\
  (major1) <  0 || \
  (major1) == 0 && (major2) <  6 || \
  (major1) == 0 && (major2) == 6 && (minor) <= 3)

/* package base-4.9.1.0 */
#define VERSION_base "4.9.1.0"
#define MIN_VERSION_base(major1,major2,minor) (\
  (major1) <  4 || \
  (major1) == 4 && (major2) <  9 || \
  (major1) == 4 && (major2) == 9 && (minor) <= 1)

/* package brick-0.18 */
#define VERSION_brick "0.18"
#define MIN_VERSION_brick(major1,major2,minor) (\
  (major1) <  0 || \
  (major1) == 0 && (major2) <  18 || \
  (major1) == 0 && (major2) == 18 && (minor) <= 0)

/* package comonad-5.0.2 */
#define VERSION_comonad "5.0.2"
#define MIN_VERSION_comonad(major1,major2,minor) (\
  (major1) <  5 || \
  (major1) == 5 && (major2) <  0 || \
  (major1) == 5 && (major2) == 0 && (minor) <= 2)

/* package matrix-0.3.5.0 */
#define VERSION_matrix "0.3.5.0"
#define MIN_VERSION_matrix(major1,major2,minor) (\
  (major1) <  0 || \
  (major1) == 0 && (major2) <  3 || \
  (major1) == 0 && (major2) == 3 && (minor) <= 5)

/* package unix-2.7.2.1 */
#define VERSION_unix "2.7.2.1"
#define MIN_VERSION_unix(major1,major2,minor) (\
  (major1) <  2 || \
  (major1) == 2 && (major2) <  7 || \
  (major1) == 2 && (major2) == 7 && (minor) <= 2)

/* tool alex-3.2.3 */
#define TOOL_VERSION_alex "3.2.3"
#define MIN_TOOL_VERSION_alex(major1,major2,minor) (\
  (major1) <  3 || \
  (major1) == 3 && (major2) <  2 || \
  (major1) == 3 && (major2) == 2 && (minor) <= 3)

/* tool gcc-7 */
#define TOOL_VERSION_gcc "7"
#define MIN_TOOL_VERSION_gcc(major1,major2,minor) (\
  (major1) <  7 || \
  (major1) == 7 && (major2) <  0 || \
  (major1) == 7 && (major2) == 0 && (minor) <= 0)

/* tool ghc-8.0.2 */
#define TOOL_VERSION_ghc "8.0.2"
#define MIN_TOOL_VERSION_ghc(major1,major2,minor) (\
  (major1) <  8 || \
  (major1) == 8 && (major2) <  0 || \
  (major1) == 8 && (major2) == 0 && (minor) <= 2)

/* tool ghc-pkg-8.0.2 */
#define TOOL_VERSION_ghc_pkg "8.0.2"
#define MIN_TOOL_VERSION_ghc_pkg(major1,major2,minor) (\
  (major1) <  8 || \
  (major1) == 8 && (major2) <  0 || \
  (major1) == 8 && (major2) == 0 && (minor) <= 2)

/* tool haddock-2.17.3 */
#define TOOL_VERSION_haddock "2.17.3"
#define MIN_TOOL_VERSION_haddock(major1,major2,minor) (\
  (major1) <  2 || \
  (major1) == 2 && (major2) <  17 || \
  (major1) == 2 && (major2) == 17 && (minor) <= 3)

/* tool happy-1.19.8 */
#define TOOL_VERSION_happy "1.19.8"
#define MIN_TOOL_VERSION_happy(major1,major2,minor) (\
  (major1) <  1 || \
  (major1) == 1 && (major2) <  19 || \
  (major1) == 1 && (major2) == 19 && (minor) <= 8)

/* tool hpc-0.67 */
#define TOOL_VERSION_hpc "0.67"
#define MIN_TOOL_VERSION_hpc(major1,major2,minor) (\
  (major1) <  0 || \
  (major1) == 0 && (major2) <  67 || \
  (major1) == 0 && (major2) == 67 && (minor) <= 0)

/* tool hsc2hs-0.68.1 */
#define TOOL_VERSION_hsc2hs "0.68.1"
#define MIN_TOOL_VERSION_hsc2hs(major1,major2,minor) (\
  (major1) <  0 || \
  (major1) == 0 && (major2) <  68 || \
  (major1) == 0 && (major2) == 68 && (minor) <= 1)

/* tool hscolour-1.24 */
#define TOOL_VERSION_hscolour "1.24"
#define MIN_TOOL_VERSION_hscolour(major1,major2,minor) (\
  (major1) <  1 || \
  (major1) == 1 && (major2) <  24 || \
  (major1) == 1 && (major2) == 24 && (minor) <= 0)

/* tool pkg-config-0.29.1 */
#define TOOL_VERSION_pkg_config "0.29.1"
#define MIN_TOOL_VERSION_pkg_config(major1,major2,minor) (\
  (major1) <  0 || \
  (major1) == 0 && (major2) <  29 || \
  (major1) == 0 && (major2) == 29 && (minor) <= 1)

/* tool strip-2.30 */
#define TOOL_VERSION_strip "2.30"
#define MIN_TOOL_VERSION_strip(major1,major2,minor) (\
  (major1) <  2 || \
  (major1) == 2 && (major2) <  30 || \
  (major1) == 2 && (major2) == 30 && (minor) <= 0)

#define CURRENT_COMPONENT_ID "caFinal-0.1.0.0"

#define CURRENT_PACKAGE_KEY "caFinal-0.1.0.0"


#undef CURRENT_PACKAGE_KEY
#undef CURRENT_COMPONENT_ID


src/CellularAutomata.hs
{-# LANGUAGE DeriveFunctor #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE AllowAmbiguousTypes #-}
{-# LANGUAGE TypeFamilies, FlexibleContexts #-}
{-# LANGUAGE TypeSynonymInstances, FlexibleInstances #-}

module CellularAutomata where

import Data.Matrix
import Control.Comonad
import System.Console.ANSI
import Control.Concurrent
import System.Random
import Control.Monad.State
import Brick (attrName, AttrName)
import Data.Vector as Vec (toList)  

sizeX::Int
sizeX = 300

sizeY::Int
sizeY = 300


type Point = (Int, Int)

data Cell = Alive | Dead deriving(Eq)

data Brian = ALV | DED | DYING deriving(Eq)

data Grid a = Grid (Matrix a) Point deriving(Show, Functor)


myRand :: State StdGen Int
myRand = do
    gen <- get
    let (r, nextGen) = randomR (1,300) gen
    put nextGen
    return r

randPair::State StdGen (Int,Int)
randPair = do
            x<-myRand
            y<-myRand
            return (x,y)

cells = take 4500 $ evalState (mapM (\_ -> randPair) $ repeat ()) $ mkStdGen 1
gun = [(3,8),(3,9),(4,8),(4,9),(13,8),(13,9),(13,10),(14,7),
         (14,8),(15,6),(15,12),(16,6),(16,12),(17,9),(18,7),
         (18,11),(19,8),(19,9),(19,10),(20,9),(23,6),(23,7),
         (23,8),(24,6),(24,7),(24,8),(25,5),(25,9),(27,4),
         (27,5),(27,9),(27,10),(37,6),(37,7),(38,6),(38,7)]  

--cells = gun ++ map (\(x,y)->(x,y+15)) gun ++ map (\(x,y)->(x,y+30)) gun



neighbors = [(x,y) | x <- [-1,0,1], y <-[-1,0,1], (x,y) /= (0,0)]

instance Show Cell where
  show Alive = "@"
  show Dead  = " "

instance Show Brian where
  show ALV = "#"
  show DED = " "
  show DYING = "1"


instance Comonad Grid where
  -- extract:: w a -> a
  extract (Grid mtx (x,y)) = getElem x y mtx
  -- extend::(w a -> b)->w a->w b
  extend f = fmap f . duplicate
  -- duplicate::w a -> w(w a) 
  duplicate (Grid mtx (x,y)) = Grid (matrix sizeX sizeY (\(w,z)->(Grid mtx (w,z)))) (x,y)


class CA w c where
  next_gen::w->c
  
  neighborValues::w->[c]
  start_state::w 
  showGrid::w->String
  cellAtt::c->AttrName
  gridRow::Int->w->[c]


getNeighbors::Grid a->[(Int, Int)]
getNeighbors (Grid mtx (x,y))=filter (\(a,b)->if a==0||b==0||a>sizeX||b>sizeY then False else True) (map (\(w,z)->(w+x, y+z)) neighbors)




instance CA (Grid Cell) Cell where
  start_state = Grid (matrix sizeX sizeY (\(x,y)->if (any ((x,y)==) cells) then Alive else Dead)) (5,1)
  next_gen (Grid mtx (x,y)) = case (getElem x y mtx) of
                                  Alive -> case (length(filter(\x->x==Alive) (neighborValues (Grid mtx (x,y))))) of
                                            2-> Alive
                                            3-> Alive
                                            _-> Dead
                                  Dead -> case (length(filter(\x->x==Alive) (neighborValues (Grid mtx (x,y))))) of 
                                            3-> Alive
                                            _-> Dead
  neighborValues (Grid mtx (x,y)) = fmap (\(y,z)->getElem y z mtx) $ getNeighbors (Grid mtx (x,y))

           
  showGrid (Grid mtx p) = show mtx
  cellAtt x = case x of
              Alive -> attrName "aliveAttr"
              Dead  -> attrName "deadAttr"
  gridRow r (Grid mtx p) = Vec.toList $ getRow r mtx
                      

instance CA (Grid Brian) Brian where
  start_state = Grid (matrix sizeX sizeY (\(x,y)->if (any ((x,y)==) cells) then ALV else DED)) (5,1)
  next_gen (Grid mtx (x,y)) = case (getElem x y mtx) of
                                      ALV -> DYING
                                      DED -> case (length(filter(\x->x==ALV) (neighborValues (Grid mtx (x,y))))) of
                                            2->ALV
                                            _->DED
                                      DYING -> DED 
  neighborValues (Grid mtx (x,y)) = fmap (\(y,z)->getElem y z mtx) $ getNeighbors (Grid mtx (x,y))

  showGrid (Grid mtx p) = show mtx
  cellAtt x = case x of
                ALV -> attrName "aliveAttr"
                DED  -> attrName"deadAttr"
                DYING -> attrName "dyingAttr"
                
                
  gridRow r (Grid mtx p) = Vec.toList $ getRow r mtx


start_state_brian::Grid Brian
start_state_brian = Grid (matrix sizeX sizeY (\(x,y)->if (any ((x,y)==) cells) then ALV else DED)) (5,1)





